### DISCUSSION TITLE
Disabled elements of a form have null values when validate

### DISCUSSION THREAD

**Discussion by ThierryM** [sqlpage/SQLPage#375] [Q&A] [answered]
Hi,
I wonder if it's a normal behavior when I submit a form where there are disabled elements, their values are "null".
Here a screen shot of my form with 3 disabled elements but that aren't empty :
![Sélection_658](https://github.com/lovasoa/SQLpage/assets/1470067/34e397d4-3c9a-4a98-94bf-736fa99a705a)
When I validate, I obtain errors with empty field like this : 
```
error returned from database: (code: 1299) NOT NULL constraint failed: besoins_eleve.eleve_id
(code: 1299) NOT NULL constraint failed: besoins_eleve.eleve_id
```
The form definition of this element is : 
` select 'eleve_id_besoin' as name, 'Elève ID' as label, 'number' as type, 1 as width, $eleve_id_besoin as value, TRUE as readonly, TRUE as disabled;`

I have to cancel "TRUE as disabled" to obtain an usable result. Is it normal ?
The disabled option is interesting because the form elements are not selectionnable, thing that's possible with the readonly option (that seems like a modifiable element).

Regards

**Comment by lovasoa** [comment]
This is the way disabled form fields are designed to work on the web, not just in SQLPage : 

> The Boolean disabled attribute, when present, makes the element not mutable, focusable, or even submitted with the form. 
> - https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/disabled#overview

And it looks like you even found what you needed to use instead yourself: the [`readonly`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly) attribute.

**Comment by ThierryM** [comment]
Thanks Ophir for the explanations.
I would be useful to have a particular looking for the readonly element to distinct easily from normal form element (I think it could be done with CSS but I don't now how to do).
Regards.

**Reply by lovasoa** [reply]
I'm not sure we want to change that... The way disabled and readonly fields display in SQLPage is just the default look for those attributes on the web.

**Reply by lovasoa** [reply, marked as answer]
What I would do in your case is move the readonly information completely out of the form, to display it in a datagrid component, and if you need it on the backend, submit it as URL parameters or hidden form fields.

**Comment by ThierryM** [comment]
Thanks for the solutions ! 'hidden' should be perfect !
I didn't know the 'hidden" parameter form form fields : it's not in the https://sql.ophir.dev/documentation.sql?component=form#component page.
Regards

**Reply by lovasoa** [reply]
Have a look at example 7 !

**Comment by ThierryM** [comment]
THANKS !!! It's exactly what I need !
